File: tree/543/543.js
var diameterOfBinaryTree = function(root) {
  if (!root) return 0;
  const lh = getHeight(root.left);
  const rh = getHeight(root.right);
  const ld = diameterOfBinaryTree(root.left);
  const rd = diameterOfBinaryTree(root.right);
  return Math.max(lh + rh, Math.max(ld, rd));
};

const getHeight = root => {
  if (!root) return 0;
  const lh = getHeight(root.left);
  const rh = getHeight(root.right);
  return 1 + Math.max(lh, rh);
};

//more compact solution
var diameterOfBinaryTree = function(root) {
  let lp = 0;

  const getHeight = root => {
    if (!root) return 0;
    const lh = getHeight(root.left);
    const rh = getHeight(root.right);
    lp = Math.max(lp, lh + rh);
    return 1 + Math.max(lh, rh);
  };
  getHeight(root);

  return lp;
};

File: tree/543/543.py
class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        def height(root):
            if root is None:
                return 0
            lh = height(root.left)
            rh = height(root.right)
            return 1 + max(lh, rh)

        if root is None:
            return 0
        lh = height(root.left)
        rh = height(root.right)
        ld = self.diameterOfBinaryTree(root.left)
        rd = self.diameterOfBinaryTree(root.right)
        return max(lh + rh, max(ld, rd))

    # more compact solution
    def diameterOfBinaryTree(self, root: TreeNode) -> int:

        self.ans = 0

        def height(root):
            if not root:
                return 0
            left = height(root.left)
            right = height(root.right)
            self.ans = max(self.ans, left + right)
            return max(left, right) + 1

        height(root)

        return self.ans

